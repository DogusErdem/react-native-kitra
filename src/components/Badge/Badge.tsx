import { Text, StyleSheet, View } from 'react-native';
import Animated, { FadeIn, FadeOut } from 'react-native-reanimated';
import type { BadgeProps, FCCWD } from '../../types';
import { applyDefaults } from '../../core/KitraProvider';

const sizes = {
  medium: {
    width: 24,
    height: 24,
    fontSize: 10,
  },
  small: {
    width: 10,
    height: 10,
    fontSize: 8,
  },
};
const badgeStyles = StyleSheet.create({
  container: {
    justifyContent: 'center', alignItems: 'center',
  },
});
const Badge : FCCWD<BadgeProps> = (
  { theme,
    visible = true,
    label,
    size = 'medium',
    containerStyle,
    textStyles,
    borderStyle = 'circular',
    icon },
) => (
  <View>
    {visible ? (
      <Animated.View
        exiting={FadeOut.duration(300)}
        entering={FadeIn.duration(400)}
        style={[badgeStyles.container,
          [borderStyle === 'circular' ? { width: size === 'medium' ? (label || icon ? 30 : 24) : 10,
            height: size === 'medium' ? (label || icon ? 30 : 24) : 10 } :
            { paddingVertical: size === 'medium' ? 4 : 3, paddingHorizontal: size === 'medium' ? 8 : 6 },
          { backgroundColor: theme?.primary, borderRadius: borderStyle === 'circular' ? 50 : 3 }],
          containerStyle]}
      >
        {(() => {
          if (label) {
            return <Text style={[{ color: theme?.white, fontSize: sizes[size]?.fontSize }, textStyles]}>{label}</Text>;
          }
          if (icon) {
            return icon;
          }
          return null;
        })()}
      </Animated.View>
    )
      : null
}
  </View>
);
export default applyDefaults(Badge);
