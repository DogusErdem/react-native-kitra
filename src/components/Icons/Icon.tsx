import { getIconType } from '../../utilities';
import { applyDefaults } from '../../core/KitraProvider';
import type { FCCWD, IconComponentProps, IconType } from '../../types';

type IconProps = {
    type: IconType,
} & IconComponentProps;

const Icon:FCCWD<IconProps> = ({ type, ...props }) => {
  const IconComponent = getIconType(type || 'ionicons');

  return (
    <IconComponent {...props} />
  );
};

export default applyDefaults(Icon);
