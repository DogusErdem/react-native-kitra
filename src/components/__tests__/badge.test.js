import { render } from '@testing-library/react-native';
import Badge from '../Badge/Badge';
import Icon from '../Icons/Icon';

/* eslint-disable no-undef */
describe('Badge', () => {
  describe('renders with visible', () => {
    const tree = render(<Badge visible />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  describe('renders badge with different sizes', () => {
    it('renders badge with small size', () => {
      const tree = render(<Badge size="small" />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders badge with medium size', () => {
      const tree = render(<Badge size="medium" />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders badge with medium size, visable prop is false', () => {
      render(<Badge size="medium" visible={false} />).toJSON();
    });
  });
  describe('renders badge with borderStyle', () => {
    it('renders badge with circular borderStyle', () => {
      const tree = render(<Badge borderStyle="circular" size="small" />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders badge with rectangular borderStyle, small size', () => {
      const tree = render(<Badge borderStyle="rectangular" size="small" />).toJSON();
      expect(tree).toMatchSnapshot();
    }); it('renders badge with rectangular borderStyle', () => {
      const tree = render(<Badge borderStyle="rectangular" size="medium" />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('renders badge with icon', () => {
    const tree = render(<Badge icon={<Icon name="check" type="ant-design" size={20} />} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  describe('renders badge with label', () => {
    const tree = render(<Badge label="Test label" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
