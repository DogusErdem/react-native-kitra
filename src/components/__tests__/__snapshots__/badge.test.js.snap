// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<View>
  <View
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    style={
      [
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
        [
          {
            "height": 24,
            "width": 24,
          },
          {
            "backgroundColor": undefined,
            "borderRadius": 50,
          },
        ],
        undefined,
      ]
    }
  />
</View>
`;

exports[` 2`] = `
<View>
  <View
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    style={
      [
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
        [
          {
            "height": 30,
            "width": 30,
          },
          {
            "backgroundColor": undefined,
            "borderRadius": 50,
          },
        ],
        undefined,
      ]
    }
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": undefined,
            "fontSize": 20,
          },
          undefined,
          {
            "fontFamily": "anticon",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      î˜²
    </Text>
  </View>
</View>
`;

exports[` 3`] = `
<View>
  <View
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    style={
      [
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
        [
          {
            "height": 30,
            "width": 30,
          },
          {
            "backgroundColor": undefined,
            "borderRadius": 50,
          },
        ],
        undefined,
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": undefined,
            "fontSize": 10,
          },
          undefined,
        ]
      }
    >
      Test label
    </Text>
  </View>
</View>
`;

exports[`Badge renders badge with borderStyle renders badge with circular borderStyle 1`] = `
<View>
  <View
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    style={
      [
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
        [
          {
            "height": 10,
            "width": 10,
          },
          {
            "backgroundColor": undefined,
            "borderRadius": 50,
          },
        ],
        undefined,
      ]
    }
  />
</View>
`;

exports[`Badge renders badge with borderStyle renders badge with rectangular borderStyle 1`] = `
<View>
  <View
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    style={
      [
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
        [
          {
            "paddingHorizontal": 8,
            "paddingVertical": 4,
          },
          {
            "backgroundColor": undefined,
            "borderRadius": 3,
          },
        ],
        undefined,
      ]
    }
  />
</View>
`;

exports[`Badge renders badge with borderStyle renders badge with rectangular borderStyle, small size 1`] = `
<View>
  <View
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    style={
      [
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
        [
          {
            "paddingHorizontal": 6,
            "paddingVertical": 3,
          },
          {
            "backgroundColor": undefined,
            "borderRadius": 3,
          },
        ],
        undefined,
      ]
    }
  />
</View>
`;

exports[`Badge renders badge with different sizes renders badge with medium size 1`] = `
<View>
  <View
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    style={
      [
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
        [
          {
            "height": 24,
            "width": 24,
          },
          {
            "backgroundColor": undefined,
            "borderRadius": 50,
          },
        ],
        undefined,
      ]
    }
  />
</View>
`;

exports[`Badge renders badge with different sizes renders badge with small size 1`] = `
<View>
  <View
    entering={BaseAnimationMock {}}
    exiting={BaseAnimationMock {}}
    style={
      [
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
        [
          {
            "height": 10,
            "width": 10,
          },
          {
            "backgroundColor": undefined,
            "borderRadius": 50,
          },
        ],
        undefined,
      ]
    }
  />
</View>
`;
